providers/common/kdfs/kbkdf.c

```diff
@@ -0,0 +1,303 @@
+    } 
+    mdname = EVP_MD_name(md);
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
+                                                 (char *)mdname, 0);
+    params[1] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
+                                                  ctx->KI, ctx->KI_len);
+    params[2] = OSSL_PARAM_construct_end();
+
+    mac = EVP_MAC_fetch(provctx, "HMAC", NULL);
+    if (mac == NULL)
+        return 0;
+    ctx_init = EVP_MAC_CTX_new(mac);
+    if (ctx_init == NULL)
+        goto done;
+
+    if (!EVP_MAC_CTX_set_params(ctx_init, params) || !EVP_MAC_init(ctx_init))
+        goto done;
+
+    h = EVP_MAC_size(ctx_init);
+    if (h == 0)
+        goto done;
+
+    K_i = OPENSSL_zalloc(h);
+    if (K_i == NULL)
+        goto done;
+
+    ret = kbkdf_derive_counter(
+        ctx_init, ctx->Label, ctx->Label_len, ctx->Context, ctx->Context_len,
+        K_i, h, L, key, keylen);
+done:
+    if (ret != 1)
+        OPENSSL_cleanse(key, keylen);
+    EVP_MAC_CTX_free(ctx_init);
+    OPENSSL_clear_free(K_i, h);
+    return ret;
+}
+
+static int kbkdf_set_buffer(unsigned char **out, size_t *out_len,
+                            const OSSL_PARAM *p)
+{
+    if (p->data == NULL || p->data_size == 0)
+        return 1;
+
+    OPENSSL_clear_free(*out, *out_len);
+    *out = NULL;
+    return OSSL_PARAM_get_octet_string(p, (void **)out, 0, out_len);
+}
+
+static int kbkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p;
+    KBKDF *ctx = vctx;
+    OPENSSL_CTX *provctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+
+    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, provctx))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_KEY);
+    if (p != NULL && !kbkdf_set_buffer(&ctx->KI, &ctx->KI_len, p))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_SALT);
+    if (p != NULL && !kbkdf_set_buffer(&ctx->Label, &ctx->Label_len, p))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_INFO);
+    if (p != NULL && !kbkdf_set_buffer(&ctx->Context, &ctx->Context_len, p))
+        return 0;
+
+    return 1;
+}
+
+static const OSSL_PARAM *kbkdf_settable_ctx_params(void)
+{
+    static const OSSL_PARAM known_settable_ctx_params[] = {
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_INFO, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_SALT, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_KEY, NULL, 0),
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_DIGEST, NULL, 0),
+
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END,
+    };
+    return known_settable_ctx_params;
+}
+
+static int kbkdf_size(KBKDF *ctx)
+{
+    int sz;
+    const EVP_MD *md;
+
+    md = ossl_prov_digest_md(&ctx->digest);
+    if (md == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_MESSAGE_DIGEST);
+        return 0;
+    }
+
+    sz = EVP_MD_size(md);
```

Comment:
This, I don't quite understand...  when reading the spec, what I see is this:

> Return: *K<sub>O</sub>, i.e., the leftmost *L* bits of *`result(n)`*. 

Ref: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf